addons/product/models/product_template.py

```diff
@@ -967,24 +1023,24 @@ def _get_possible_combinations(self, parent_combination=None, necessary_values=N
 
         necessary_values = necessary_values or self.env['product.template.attribute.value']
         necessary_attributes = necessary_values.mapped('attribute_id')
-        ptal_stack = [self.valid_product_template_attribute_line_ids.filtered(lambda ptal: ptal.attribute_id not in necessary_attributes)]
-        combination_stack = [necessary_values]
-
-        # keep going while we have attribute lines to test
-        while len(ptal_stack):
-            attribute_lines = ptal_stack.pop()
-            combination = combination_stack.pop()
-
-            if not attribute_lines:
-                # full combination, if it's possible return it, otherwise skip it
-                if self._is_combination_possible(combination, parent_combination):
-                    yield(combination)
-            else:
-                # we have remaining attribute lines to consider
-                for ptav in reversed(attribute_lines[0].product_template_value_ids):
-                    ptal_stack.append(attribute_lines[1:])
-                    combination_stack.append(combination + ptav)
 
+        attributes = self.valid_product_template_attribute_line_ids.filtered(lambda ptal: ptal.attribute_id not in necessary_attributes)
+
+        if not attributes and self._is_combination_possible(necessary_values, parent_combination):
+            yield necessary_values
+
+        # compute all possible values for each attribute
+        attributes_values = [
+            attribute.product_template_value_ids
+            for attribute in attributes
+        ]
+
+        for product in self._cartesian_product(attributes_values):
+            combination = sum(product, necessary_values)
+            if self._is_combination_possible(combination, parent_combination):
+                yield combination
+
+        # Is this required ? A generator will raise StopIteration by default when it's empty and polled
```

Comment:
Well you used a generator inside a generator so you basically "moved" the error into the other method, but it's better to keep this intact and remove the other method as suggested above.

Reply:
No, I didn't move that part of the logic, it already worked that way before. This code is reached when the `_get_possible_combinations` generator is polled and the for loop above is exhausted. This for loop already existed and worked the same way ; some iterations are skipped inside the new method rather than inside the loop, that doesn't change anything for the code after the loop.

The comment was expressing my surprise that we break the iteration protocol. After digging more, I've seen that it has been introduced by a fix to avoid silent errors where the caller wouldn't handle the `StopIteration` error.